govcd/vm.go

```diff
@@ -850,3 +851,186 @@ func (vm *VM) SetGuestCustomizationSection(guestCustomizationSection *types.Gues
+	}
+
+	err = vm.validateInternalDiskInput(diskData)
+	if err != nil {
+		return "", err
+	}
+
+	var diskSettings []*types.DiskSettings
+	if vm.VM.VmSpecSection != nil && vm.VM.VmSpecSection.DiskSection != nil && vm.VM.VmSpecSection.DiskSection.DiskSettings != nil {
+		diskSettings = vm.VM.VmSpecSection.DiskSection.DiskSettings
+	}
+
+	diskSettings = append(diskSettings, diskData)
+	vmSpecSection := vm.VM.VmSpecSection
+	vmSpecSection.DiskSection.DiskSettings = diskSettings
+
+	vmSpecSection, err = vm.UpdateInternalDisks(vmSpecSection)
+	if err != nil {
+		return "", err
+	}
+
+	for _, diskSetting := range vmSpecSection.DiskSection.DiskSettings {
+		if diskSetting.AdapterType == diskData.AdapterType &&
+			diskSetting.BusNumber == diskData.BusNumber &&
+			diskSetting.UnitNumber == diskData.UnitNumber {
+			return diskSetting.DiskId, nil
+		}
+	}
+
+	return "", fmt.Errorf("created disk wasn't in list of returned VM internal disks")
+}
+
+func (vm *VM) validateInternalDiskInput(diskData *types.DiskSettings) error {
+	if diskData.AdapterType == "" {
+		return fmt.Errorf("disk settings missing required field: adapter type")
+	}
+
+	if diskData.BusNumber < 0 {
+		return fmt.Errorf("disk settings bus number has to be 0 or higher")
+	}
+
+	if diskData.UnitNumber < 0 {
+		return fmt.Errorf("disk settings unit number has to be 0 or higher")
+	}
+
+	if diskData.SizeMb < int64(0) {
+		return fmt.Errorf("disk settings size MB has to be 0 or higher")
+	}
+
+	if diskData.Iops != nil && *diskData.Iops < int64(0) {
+		return fmt.Errorf("disk settings iops has to be 0 or higher")
+	}
+
+	if diskData.ThinProvisioned == nil {
+		return fmt.Errorf("disk settings missing required field: thin provisioned")
+	}
+
+	if diskData.StorageProfile == nil {
+		return fmt.Errorf("disk settings missing required field: storage profile")
+	}
+
+	return nil
+}
+
+// GetInternalDiskById returns a valid *types.DiskSettings if it exists.
+// If it doesn't, returns nil and ErrorEntityNotFound or other err.
+func (vm *VM) GetInternalDiskById(diskId string, refresh bool) (*types.DiskSettings, error) {
+	if refresh {
+		err := vm.Refresh()
+		if err != nil {
+			return nil, fmt.Errorf("error refreshing VM: %s", err)
+		}
+	}
+
+	if diskId == "" {
+		return nil, fmt.Errorf("cannot get internal disk - provided disk Id is empty")
+	}
+
+	if vm.VM.VmSpecSection.DiskSection == nil || vm.VM.VmSpecSection.DiskSection.DiskSettings == nil ||
+		len(vm.VM.VmSpecSection.DiskSection.DiskSettings) == 0 {
+		return nil, fmt.Errorf("cannot get internal disk - VM doesn't have internal disks")
+	}
+
+	for _, diskSetting := range vm.VM.VmSpecSection.DiskSection.DiskSettings {
+		if diskSetting.DiskId == diskId {
+			return diskSetting, nil
+		}
+	}
+
+	return nil, ErrorEntityNotFound
+}
+
+// DeleteInternalDiskById delete disk using provided disk ID.
+// Runs synchronously, VM is ready for another operation after this function returns.
+func (vm *VM) DeleteInternalDiskById(diskId string) error {
```

Comment:
Why is the `ById` in the name? We don't have `DeleteInternalDiskByName`, and BTW, the changelog has the name without the `ById` suffix

Reply:
This was request by Dainius. I am also fine with it. Leave more abstract name for feature in case we may need it. FYI - update changelog. Thank you for noticing

Reply:
The var `diskId` confused me. If it is ID then I thought it is ById

Reply:
Probably the confusion comes from a different approach we're taking here. 
In other entities, we do `entity.Delete`, while here we do `parent.DeleteEntityById`.
In both case, I think the `ById` suffix is not needed, as it suggests that there are other `DeleteBySomeField` which we don't have.

Reply:
@Didainius is it fine I revert back?

Reply:
Go ahead - it is not a critical thing although to my mind this name is closer to reality :)

Reply:
done